libapicore/ApiServer.cpp

```diff
@@ -423,6 +424,13 @@ Json::Value ApiConnection::getMinerStatHR()
 		gpuIndex++;
 	}
 
+	gpuIndex = 0;
```

Comment:
Instead of having 3 different for loops (inlcuding this last yours) can we manage to pull info from all gpus using only 1 for loop ?

Reply:
Also ... ispaused is a little too generic. 
Maybe we could add some info about the pausing reason ? (eg. temp threshold )

Reply:
loops :+1: 

ispaused: I already thought about that and came due following reasons to the current solution:
* as I implemented -tstop/-tstart I already thought about a pause bit flag (TEMPERATURE, USER, ...)
but currently we have only pausing due temperature reason (no keypress and no api to pause mining)
* ispaused is a bool, if there will be more reasons I thought the API can respond a "pausereasons" or so. I've noticed content strings changes more often than bools ;-) which always hurts in interfaces.

So I have following questions before adapting PR:
* ispaused: bool or string
* If ispaused is a string: when not paused: null or "" or "no" or "0" ?
* If ispaused is a string and if multiple reasons for pausing pending - join reasons with "," or ";"
* Should I prepare pausing to be a unsigned int with bits and adapt is_mining_paused() ?

Reply:
I'll try explain what I would like to achieve.
At present the result of getstathr is like this

```
{  
   "id":0,
   "jsonrpc":"2.0",
   "result":{  
      "ethhashrate":87388615,
      "ethhashrates":[  
         14506510,
         14593898,
         14593898,
         14593898,
         14506510,
         14593898
      ],
      "ethinvalid":0,
      "ethpoolsw":0,
      "ethrejected":0,
      "ethshares":136,
      "fanpercentages":[  
         90,
         90,
         90,
         90,
         90,
         90
      ],
      "pooladdrs":"10.0.0.113:5551",
      "powerusages":[  
         0.0,
         0.0,
         0.0,
         0.0,
         0.0,
         0.0
      ],
      "runtime":"112",
      "temperatures":[  
         54,
         51,
         57,
         59,
         62,
         60
      ],
      "version":"0.15.0.dev11-96+commit.6b02ef2f"
   }
}
```

I whish we could manage gpu's as an array of objects (instead of parsing a string which has to be split into an array of values)

```
{  
   "id":0,
   "jsonrpc":"2.0",
   "result":{  
      "hashrate":87388615,
      "shares":136,
      "invalid":0,
	  "rejected":0,
      "poolsw":0,
      "pooladdrs":"10.0.0.113:5551",
      "runtime":"112",
      "version":"0.15.0.dev11-96+commit.6b02ef2f.dirty",
	  "gpus" : [
			{ "index" : 0,
			  "hashrate : 14506510,
			  "power" : 0.0,
			  "fan" : 90,
			  "temp" : 54,
			  "temp-stop" : 0,
			  "temp-start" : 0,
			  "status" : "<operational|paused|error>",
			  "reason" : "<null|at-temp|manual|error>"
			},
			{ "index" : 1,
			  "hashrate : 14506510,
			  "power" : 0.0,
			  "fan" : 90,
			  "temp" : 54,
			  "temp-stop" : 0,
			  "temp-start" : 0,
			  "status" : "<operational|paused|error>",
			  "reason" : "<null|at-temp|manual|error>"
			},
			{ "index" : 2,
			  "hashrate : 14506510,
			  "power" : 0.0,
			  "fan" : 90,
			  "temp" : 54
			  "temp-stop" : 0,
			  "temp-start" : 0,
			  "status" : "<operational|paused|error>",
			  "reason" : "<null|at-temp|manual|error>"
			},
			{ "index" : 3,
			  "hashrate : 14506510,
			  "power" : 0.0,
			  "fan" : 90,
			  "temp" : 54
			  "temp-stop" : 0,
			  "temp-start" : 0,
			  "status" : "<operational|paused|error>",
			  "reason" : "<null|at-temp|manual|error>"
			},
			{ "index" : 4,
			  "hashrate : 14506510,
			  "power" : 0.0,
			  "fan" : 90,
			  "temp" : 54
			  "temp-stop" : 0,
			  "temp-start" : 0,
			  "status" : "<operational|paused|error>",
			  "reason" : "<null|at-temp|manual|error>"
			},
			{ "index" : 5,
			  "hashrate : 14506510,
			  "power" : 0.0,
			  "fan" : 90,
			  "temp" : 54
			  "temp-stop" : 0,
			  "temp-start" : 0,
			  "status" : "<operational|paused|error>",
			  "reason" : "<null|at-temp|manual|error>"
			}
	  ]
	  
   }
}
```

Would not set bit-shifting values for reasons of paused as this would imply consumers of api to have a legend of values and also bit shifting may be affected by endianess.

I know this scheme is a bit wordy but is more Json style than the actual.

Obviously this is my view, not necessarily the same of other devs. Thus some more feedback may be needed.

P.s. Anticipating your question about compatibility with ... erm ... other miners : getstathr is ethminer's only ... so we can change almost everything leaving gestat1 as is.

Reply:
In this way we could also each gpu object any property we may need (ex. Vendor, Compute level and so on) changing only the structure of the GPU object while leaving the "result" structure intact.

Reply:
> Would not set bit-shifting values for reasons of paused as this would imply consumers of api to have a legend of values and also bit shifting may be affected by endianess.

I just thought preparing internal a flag for pausing (replace 'm_wait_for_tstart_temp' with 'mining_paused_flag' and adding MINIG_PAUSED_FLAG_WAIT_TSTART_TEMP=0x01, MINIG_PAUSED_FLAG_MANUAL = 0x02,  MINIG_PAUSED_FLAG_ERROR = 0x04, ...)

> P.s. Anticipating your question about compatibility with ... erm ... other miners : getstathr is ethminer's only ... so we can change almost everything leaving gestat1 as is.

Yes I know - In miner_getstathr I feel free to change nearly everything ;-)

>Obviously this is my view, not necessarily the same of other devs. Thus some more feedback may be needed.

Yep - I'll wait for more reviews...